src/DurableTask.Netherite/StorageLayer/Faster/StoreWorker.cs

```diff
@@ -349,7 +371,126 @@ void ScheduleNextIdleCheckpointTime()
             var actual = (((earliest - offset - 1) / period) + 1) * period + offset;
             this.timeOfNextIdleCheckpoint = new DateTime(actual, DateTimeKind.Utc);
         }
-        
+
+        void StartCheckpointOrFailOnTimeout(Func<bool> checkpointRoutine, string messageOnError)
+        {
+            var checkpointStarted = checkpointRoutine.Invoke();
+            if (checkpointStarted)
+            {
+                this.startFasterCheckpointRefusal = null; 
+            }
+            else
+            {
+                // track start of FASTER refusal to start checkpoint
+                var currentTime = DateTime.UtcNow;
+                this.startFasterCheckpointRefusal ??= currentTime;
+
+                // if the refusal to checkpoint started over a minute ago, terminate partition
+                TimeSpan durationFasterRefusal = currentTime - this.startFasterCheckpointRefusal.Value;
+                if (durationFasterRefusal > TimeSpan.FromMinutes(1))
+                {
+                    messageOnError += $" FASTER first refused to checkpoint at '{this.startFasterCheckpointRefusal}'. Duration of refusal = {durationFasterRefusal}";
+                    Exception err = new Exception(messageOnError);
+                    this.partition.ErrorHandler.HandleError(nameof(StartCheckpointOrFailOnTimeout), messageOnError, err, terminatePartition: true, reportAsWarning: false);
+                }
+            }
+        }
+
+        async Task RunCheckpointingStateMachine()
```

Comment:
You can make this a ValueTask to save an allocation on the common path (where the method completes synchronously)

Reply:
Since this is a rather large method, I would  consider making it an actual separate method instead of a local function. But only if it does not require you to pass a lot of context parameters.